### PR TITLE
Add offer_deferred status

### PR DESCRIPTION
**Pull request by AgaDufrat**
## Context

Providers need to be able to defer offers to the next recruitment cycle.

## Changes proposed in this pull request

- Add a new `offer_deferred` status to the Application Choice
- Add rules to the state machine
  - Only applications that are in 'Offer accepted' (`pending_conditions`) and 'Conditions met' (`recruited`) can change to `offer_deferred`
  - From `offer_deferred` we can go back to `recruited` and `pending_conditions` states and `withdrawn` as candidates should be able to withdraw 
- Update status tag components
- Update support documentation
- Update performance statistics
- Hide deferred applications from the API

## Guidance to review

- Check preview for each interface eg: http://localhost:3000/rails/view_components/provider_interface/application_status_tag_component/offer_deferred
- Check provider and candidate application flows: http://localhost:3000/support/provider-flow

- Not sure about the way of hiding deferred offers from the API.  

## Link to Trello card

https://trello.com/c/UeS5yjDh/2689-add-offerdeferred-status

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/services/application_state_change.rb [hunk 1]
```diff
@@ -5,8 +5,8 @@ class ApplicationStateChange
   # as they do not need references or the 7-day cooling off period
   STATES_THAT_MAY_BE_SENT_TO_PROVIDER = %i[application_complete unsubmitted].freeze
   STATES_NOT_VISIBLE_TO_PROVIDER = %i[unsubmitted awaiting_references application_complete cancelled].freeze
-  STATES_VISIBLE_TO_PROVIDER = %i[awaiting_provider_decision offer pending_conditions recruited rejected declined withdrawn conditions_not_met offer_withdrawn rejected_at_end_of_cycle].freeze
-  ACCEPTED_STATES = %i[pending_conditions conditions_not_met recruited].freeze
+  STATES_VISIBLE_TO_PROVIDER = %i[awaiting_provider_decision offer pending_conditions recruited rejected declined withdrawn conditions_not_met offer_withdrawn rejected_at_end_of_cycle offer_deferred].freeze
+  ACCEPTED_STATES = %i[pending_conditions conditions_not_met recruited offer_deferred].freeze
```
Comment (tijmenb) [inline app/services/application_state_change.rb:9]: Should this go into `OFFERED_STATES` instead? cc @duncanjbrown as it impacts TAD reporting.

File: app/queries/get_application_choices_for_providers.rb [hunk 2]
```diff
@@ -20,6 +20,6 @@ def self.call(providers:)
       .where('courses.provider_id' => providers, 'courses.recruitment_cycle_year' => RecruitmentCycle.visible_years)
       .or(ApplicationChoice.includes(*includes)
         .where('courses.accredited_provider_id' => providers, 'courses.recruitment_cycle_year' => RecruitmentCycle.visible_years))
-      .where('status IN (?)', ApplicationStateChange.states_visible_to_provider).includes([:accredited_provider])
+      .where('status IN (?)', ApplicationStateChange.states_visible_to_provider_in_vendor_api).includes([:accredited_provider])
```
Comment (tijmenb) [inline app/queries/get_application_choices_for_providers.rb:23]: Out of interest, what's the reason for not showing deferred offers via the API?

Comment (duncanjbrown) [reply app/queries/get_application_choices_for_providers.rb:23]: It's an unexpected state, so potentially a breaking change

File: app/services/application_state_change.rb (hunk 1 above)
Comment (duncanjbrown) [reply app/services/application_state_change.rb:9]: In fact this _is_ an accepted state bc it's not possible to enter it without first accepting the offer. However I take your point that TAD might not want to count it. Propose leaving it in here and either defining another constant below (eg `ACCEPTED_STATES_FOR_TAD`) or doing some work in the TAD export to filter it out. Think I prefer the latter?

File: app/queries/get_application_choices_for_providers.rb (hunk 2 above)
Comment (duncanjbrown) [inline app/queries/get_application_choices_for_providers.rb:23]: Will this not also hide these applications in Manage? https://github.com/DFE-Digital/apply-for-teacher-training/blob/ceeb842702cd47f3ae94b49cb9194f422c8cf060/app/controllers/provider_interface/application_choices_controller.rb#L12-L14

File: app/services/application_state_change.rb (hunk 1 above)
Comment (AgaDufrat) [reply app/services/application_state_change.rb:9]: Thanks, I will have a go at filtering it out of the TAD export.

File: app/queries/get_application_choices_for_providers.rb (hunk 2 above)
Comment (AgaDufrat) [reply app/queries/get_application_choices_for_providers.rb:23]: ah, yes! I will. Thanks!

**Comment by duncanjbrown** [comment]
It's possible to crash the app by putting an application in `offer_deferred` state and trying to visit it. Fixing.

Screenshot 2020-09-07 at 17 00 09

**Review by duncanjbrown — APPROVED** [review]
props @AgaDufrat  — merging in your absence!

**Pull request merged**